src/Nncase.Quantization/Quantization/QuantUtility.cs

```diff
@@ -0,0 +1,344 @@
+﻿// Copyright (c) Canaan Inc. All rights reserved.
+// Licensed under the Apache license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading.Tasks;
+using Nncase.IR;
+using OrtKISharp;
+using SMath = System.Math;
+
+namespace Nncase.Quantization;
+
+/// <summary>
+/// Array utility.
+/// </summary>
+public static class QuantAlgorithmUtility
+{
+    public static Span<float> SquantWeights(Span<float> inputWeights, Tensor<float> inputWeightsRanges, ReadOnlySpan<int> inputWeightsShape, QuantMode quantMode, int bits, bool isByChannel)
+    {
+        float qMax, qMin;
+        if (quantMode == QuantMode.UnsignedMode)
+        {
+            qMax = (1 << bits) - 1;
+            qMin = 0;
+        }
+        else if (quantMode == QuantMode.SignedAsymmetricMode)
+        {
+            qMax = (1 << (bits - 1)) - 1;
+            qMin = -(1 << (bits - 1));
+        }
+        else
+        {
+            qMax = (1 << (bits - 1)) - 1;
+            qMin = -(1 << (bits - 1)) + 1;
+        }
+
+        OrtKISharp.Tensor x, delta, zeroPoint;
+        if (inputWeightsShape.Length == 4)
+        {
+            var outChannel = inputWeightsShape[0];
+            var inChannel = inputWeightsShape[1];
+            var filterH = inputWeightsShape[2];
+            var filterW = inputWeightsShape[3];
+            unsafe
+            {
+                fixed (void* numPtr1 = &inputWeights.GetPinnableReference())
```

Comment:
Use `Tensor.PinBuffer` instead of `fixed` block. The pointer will be invalid out of the `fixed` block.